File: 102.binary-tree-level-order-traversal.py
#
# @lc app=leetcode id=102 lang=python3
#
# [102] Binary Tree Level Order Traversal
#
# https://leetcode.com/problems/binary-tree-level-order-traversal/description/
#
# algorithms
# Medium (51.97%)
# Likes:    2286
# Dislikes: 62
# Total Accepted:    514.9K
# Total Submissions: 988.2K
# Testcase Example:  '[3,9,20,null,null,15,7]'
#
# Given a binary tree, return the level order traversal of its nodes' values.
# (ie, from left to right, level by level).
#
#
# For example:
# Given binary tree [3,9,20,null,null,15,7],
#
# ⁠   3
# ⁠  / \
# ⁠ 9  20
# ⁠   /  \
# ⁠  15   7
#
#
#
# return its level order traversal as:
#
# [
# ⁠ [3],
# ⁠ [9,20],
# ⁠ [15,7]
# ]
#
#
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        stack = []
        nums = []
        if not root:
            return nums
        stack.append([root])
        while stack:
            nodes = stack.pop()
            children = []
            vals = []
            for node in nodes:
                vals.append(node.val)
                if node.left:
                    children.append(node.left)
                if node.right:
                    children.append(node.right)
            if children:
                stack.append(children)
            if vals:
                nums.append(vals)
        return nums


# @lc code=end

File: 102.binary-tree-level-order-traversal.cpp
/*
 * @lc app=leetcode id=102 lang=cpp
 *
 * [102] Binary Tree Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
//#include <vector>
//using std::vector;
//#include <queue>
//using std::queue;
//
//struct TreeNode {
//    int val;
//    TreeNode *left;
//    TreeNode *right;
//    TreeNode() : val(0), left(nullptr), right(nullptr) {}
//    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
//};
class Solution {
public:
    vector<vector<int> > levelOrder(TreeNode* root) {
        vector<TreeNode*> curLevel;
        vector<TreeNode*> nextLevel;
        vector<vector<int> > res;
        if(root) curLevel.push_back(root);
        while (!curLevel.empty()) {
            vector<int> curRes;
            for (vector<TreeNode*>::iterator itr = curLevel.begin(); itr != curLevel.end(); ++itr) {
                curRes.push_back((*itr)->val);
                if ((*itr)->left) nextLevel.push_back((*itr)->left);
                if ((*itr)->right) nextLevel.push_back((*itr)->right);
            }
            res.push_back(curRes);
            curRes.clear();
            curLevel = nextLevel;
            nextLevel.clear();
        } 
        return res;
    }
};
// @lc code=end

